File: templates/strings/eertreeArray.py
from bisect import bisect_left

# Palindromic tree (eertree) over an array a of any comparable, hashable values
# (ints of any size, negatives, floats, strings, tuples -- values are compared
# only for equality, never used as indices, so anything sortable works).
# Internally each distinct palindromic subarray of a is one node, but no method
# below exposes nodes -- everything is stated in terms of indices into a.
# An array of length n has at most n distinct palindromic subarrays, so every
# list returned below has size <= n.
# numUniqueValsInArray below = how many DISTINCT values actually appear in a
# (derived from a): 2 for a binary array, 4 for DNA, up to n in general.

class Eertree:
    # O(n log numUniqueValsInArray) time, O(n) memory
    # (the transition table is dense when small enough to afford, else hashed,
    # and is discarded when the build ends -- it is not kept on the object)
    def __init__(self, a):
        self.a = a
        n = self.n = len(a)
        vals = sorted(set(a))
        k = self.numUniqueValsInArray = max(1, len(vals))
        code = [bisect_left(vals, x) for x in a]
        self.palLen = [-1, 0]
        self.palLink = [0, 0]
        self.depth = [0, 0]
        self.firstEnd = [-1, -1]
        self.diff = [0, 0]
        self.sLink = [0, 0]
        self.sufNode = [0] * n
        self.newAt = [0] * n
        self._occ = None
        self._revLen = None
        self._prefCnt = None

        if (n + 2) * k <= 2_000_000:
            table = [0] * ((n + 2) * k)
            def getT(v, c):
                return table[v * k + c]
        else:
            table = {}
            def getT(v, c):
                return table.get(v * k + c, 0)
        def setT(v, c, u):
            table[v * k + c] = u

        last = 1
        for i in range(n):
            c = code[i]
            cur = last
            while True:
                j = i - self.palLen[cur] - 1
                if j >= 0 and code[j] == c:
                    break
                cur = self.palLink[cur]
            ex = getT(cur, c)
            if ex:
                last = ex
                self.sufNode[i] = last
                continue
            v = len(self.palLen)
            self.palLen.append(self.palLen[cur] + 2)
            self.firstEnd.append(i)
            if self.palLen[v] == 1:
                link = 1
            else:
                t = self.palLink[cur]
                while True:
                    j = i - self.palLen[t] - 1
                    if j >= 0 and code[j] == c:
                        break
                    t = self.palLink[t]
                link = getT(t, c)
            self.palLink.append(link)
            self.depth.append(self.depth[link] + 1)
            d = self.palLen[v] - self.palLen[link]
            self.diff.append(d)
            self.sLink.append(link if d != self.diff[link] else self.sLink[link])
            setT(cur, c, v)
            self.newAt[i] = 1
            last = v
            self.sufNode[i] = last

    # O(1) -- the smallest l such that a[l..r] is a palindrome
    def leftmostLForPalEndingAt(self, r):
        return r - self.palLen[self.sufNode[r]] + 1

    # O(n) -- list of size n, res[r] = leftmostLForPalEndingAt(r)
    def leftmostLForPalEndingAtEach(self):
        return [self.leftmostLForPalEndingAt(r) for r in range(self.n)]

    # O(n log numUniqueValsInArray) on first call (builds the eertree of reversed
    # a), O(1) after -- the largest r such that a[l..r] is a palindrome
    def rightmostRForPalStartingAt(self, l):
        if self._revLen is None:
            rev = Eertree(self.a[::-1])
            self._revLen = [rev.palLen[v] for v in rev.sufNode]
        return l + self._revLen[self.n - 1 - l] - 1

    # O(n log numUniqueValsInArray) on first call, O(n) after
    # list of size n, res[l] = rightmostRForPalStartingAt(l)
    def rightmostRForPalStartingAtEach(self):
        self.rightmostRForPalStartingAt(0)
        n = self.n
        return [l + self._revLen[n - 1 - l] - 1 for l in range(n)]

    # O(1) -- how many distinct palindromic subarrays a[0..r] has
    def numDistinctPalsInPrefix(self, r):
        if self._prefCnt is None:
            self._prefCnt = []
            run = 0
            for i in range(self.n):
                run += self.newAt[i]
                self._prefCnt.append(run)
        return self._prefCnt[r]

    # O(n) -- list of size n, res[r] = numDistinctPalsInPrefix(r)
    def numDistinctPalsInPrefixEach(self):
        self.numDistinctPalsInPrefix(0)
        return list(self._prefCnt)

    # O(1) -- how many distinct palindromic subarrays a has, at most n
    def numDistinctPals(self):
        return len(self.palLen) - 2

    # O(1) -- how many palindromic subarrays end exactly at r, counting every
    # distinct palindrome that ends there (they are all suffixes of a[0..r])
    def numPalsEndingAt(self, r):
        return self.depth[self.sufNode[r]]

    # O(n) -- list of size n, res[r] = numPalsEndingAt(r)
    def numPalsEndingAtEach(self):
        return [self.depth[self.sufNode[r]] for r in range(self.n)]

    # O(n) -- total number of palindromic subarrays of a with multiplicity
    # (so [5, 5] counts 3: two single 5's and one [5, 5]), up to n * (n + 1) / 2
    def numPalsTotal(self):
        return sum(self._occCounts()[2:])

    # O(n) -- list of (l, r, count), one entry per DISTINCT palindromic subarray
    # of a, so at most n entries. a[l..r] is that palindrome (using its first
    # occurrence) and count is how many times it occurs in a.
    def distinctPalsWithCounts(self):
        occ = self._occCounts()
        res = []
        for v in range(2, len(self.palLen)):
            e = self.firstEnd[v]
            res.append((e - self.palLen[v] + 1, e, occ[v]))
        return res

    # O(n log n) -- fewest palindromes a can be cut into
    def minPalPartition(self):
        n = self.n
        INF = float("inf")
        dp = [INF] * (n + 1)
        dp[0] = 0
        seriesAns = [0] * len(self.palLen)
        for i in range(1, n + 1):
            v = self.sufNode[i - 1]
            while self.palLen[v] > 0:
                seriesAns[v] = dp[i - self.palLen[self.sLink[v]] - self.diff[v]]
                if self.diff[v] == self.diff[self.palLink[v]]:
                    seriesAns[v] = min(seriesAns[v], seriesAns[self.palLink[v]])
                if seriesAns[v] + 1 < dp[i]:
                    dp[i] = seriesAns[v] + 1
                v = self.sLink[v]
        return dp[n]

    def _occCounts(self):
        if self._occ is None:
            occ = [0] * len(self.palLen)
            for v in self.sufNode:
                occ[v] += 1
            for v in range(len(self.palLen) - 1, 1, -1):
                occ[self.palLink[v]] += occ[v]
            occ[0] = occ[1] = 0
            self._occ = occ
        return self._occ

File: templates/strings/eertree/eertreeArray.cpp
#include <bits/stdc++.h>
using namespace std;

// Palindromic tree (eertree) over an array a of any comparable type T
// (int, long long, negatives, double, string, pair -- values are compared only
// for equality, never used as indices, so anything with operator< works).
// Internally each distinct palindromic subarray of a is one node, but no method
// below exposes nodes -- everything is stated in terms of indices into a.
// An array of length n has at most n distinct palindromic subarrays, so every
// vector returned below has size <= n.
// numUniqueValsInArray below = how many DISTINCT values actually appear in a
// (derived from a): 2 for a binary array, 4 for DNA, up to n in general.

template <class T>
struct Eertree {
    int n, numUniqueValsInArray;
    vector<T> a;
    vector<int> code;
    vector<int> palLen, palLink, depth, firstEnd, diff, sLink, sufNode, newAt;
    vector<int> occCache, prefCache, revLenCache;

    // O(n log numUniqueValsInArray) time, O(n) memory
    // (the transition table is dense when small enough to afford, else hashed,
    // and is discarded when the build ends -- it is not kept on the object)
    Eertree(const vector<T> &arr) {
        a = arr;
        n = (int)a.size();
        vector<T> vals = a;
        sort(vals.begin(), vals.end());
        vals.erase(unique(vals.begin(), vals.end()), vals.end());
        int k = numUniqueValsInArray = max<int>(1, (int)vals.size());
        code.resize(n);
        for (int i = 0; i < n; i++)
            code[i] = (int)(lower_bound(vals.begin(), vals.end(), a[i]) - vals.begin());

        palLen = {-1, 0};
        palLink = {0, 0};
        depth = {0, 0};
        firstEnd = {-1, -1};
        diff = {0, 0};
        sLink = {0, 0};
        sufNode.assign(n, 0);
        newAt.assign(n, 0);

        bool dense = (long long)(n + 2) * k <= 8000000LL;
        vector<int> tableDense;
        unordered_map<long long, int> tableSparse;
        if (dense) tableDense.assign((size_t)(n + 2) * k, 0);
        else tableSparse.reserve(2 * n + 4);

        auto getT = [&](int v, int c) -> int {
            long long idx = (long long)v * k + c;
            if (dense) return tableDense[idx];
            auto it = tableSparse.find(idx);
            return it == tableSparse.end() ? 0 : it->second;
        };
        auto setT = [&](int v, int c, int u) {
            long long idx = (long long)v * k + c;
            if (dense) tableDense[idx] = u;
            else tableSparse[idx] = u;
        };

        int last = 1;
        for (int i = 0; i < n; i++) {
            int c = code[i];
            int cur = last;
            while (true) {
                int j = i - palLen[cur] - 1;
                if (j >= 0 && code[j] == c) break;
                cur = palLink[cur];
            }
            int ex = getT(cur, c);
            if (ex) { last = ex; sufNode[i] = last; continue; }
            int v = (int)palLen.size();
            palLen.push_back(palLen[cur] + 2);
            firstEnd.push_back(i);
            int link;
            if (palLen[v] == 1) link = 1;
            else {
                int t = palLink[cur];
                while (true) {
                    int j = i - palLen[t] - 1;
                    if (j >= 0 && code[j] == c) break;
                    t = palLink[t];
                }
                link = getT(t, c);
            }
            palLink.push_back(link);
            depth.push_back(depth[link] + 1);
            int d = palLen[v] - palLen[link];
            diff.push_back(d);
            sLink.push_back(d != diff[link] ? link : sLink[link]);
            setT(cur, c, v);
            newAt[i] = 1;
            last = v;
            sufNode[i] = last;
        }
    }

    // O(1) -- the smallest l such that a[l..r] is a palindrome
    int leftmostLForPalEndingAt(int r) {
        return r - palLen[sufNode[r]] + 1;
    }

    // O(n) -- vector of size n, res[r] = leftmostLForPalEndingAt(r)
    vector<int> leftmostLForPalEndingAtEach() {
        vector<int> res(n);
        for (int r = 0; r < n; r++) res[r] = leftmostLForPalEndingAt(r);
        return res;
    }

    // O(n log numUniqueValsInArray) on first call (builds the eertree of reversed
    // a), O(1) after -- the largest r such that a[l..r] is a palindrome
    int rightmostRForPalStartingAt(int l) {
        if (revLenCache.empty()) {
            vector<T> rev(a.rbegin(), a.rend());
            Eertree<T> t(rev);
            revLenCache.resize(n);
            for (int i = 0; i < n; i++) revLenCache[i] = t.palLen[t.sufNode[i]];
        }
        return l + revLenCache[n - 1 - l] - 1;
    }

    // O(n log numUniqueValsInArray) on first call, O(n) after
    // vector of size n, res[l] = rightmostRForPalStartingAt(l)
    vector<int> rightmostRForPalStartingAtEach() {
        vector<int> res(n);
        for (int l = 0; l < n; l++) res[l] = rightmostRForPalStartingAt(l);
        return res;
    }

    // O(1) -- how many distinct palindromic subarrays a[0..r] has
    int numDistinctPalsInPrefix(int r) {
        if (prefCache.empty()) {
            prefCache.resize(n);
            int run = 0;
            for (int i = 0; i < n; i++) { run += newAt[i]; prefCache[i] = run; }
        }
        return prefCache[r];
    }

    // O(n) -- vector of size n, res[r] = numDistinctPalsInPrefix(r)
    vector<int> numDistinctPalsInPrefixEach() {
        numDistinctPalsInPrefix(0);
        return prefCache;
    }

    // O(1) -- how many distinct palindromic subarrays a has, at most n
    int numDistinctPals() {
        return (int)palLen.size() - 2;
    }

    // O(1) -- how many palindromic subarrays end exactly at r, counting every
    // distinct palindrome that ends there (they are all suffixes of a[0..r])
    int numPalsEndingAt(int r) {
        return depth[sufNode[r]];
    }

    // O(n) -- vector of size n, res[r] = numPalsEndingAt(r)
    vector<int> numPalsEndingAtEach() {
        vector<int> res(n);
        for (int r = 0; r < n; r++) res[r] = depth[sufNode[r]];
        return res;
    }

    // O(n) -- total number of palindromic subarrays of a with multiplicity
    // (so [5, 5] counts 3: two single 5's and one [5, 5]), up to n * (n + 1) / 2
    long long numPalsTotal() {
        occCounts();
        long long tot = 0;
        for (int v = 2; v < (int)occCache.size(); v++) tot += occCache[v];
        return tot;
    }

    // O(n) -- vector of (l, r, count), one entry per DISTINCT palindromic
    // subarray of a, so at most n entries. a[l..r] is that palindrome (using its
    // first occurrence) and count is how many times it occurs in a.
    vector<array<int, 3>> distinctPalsWithCounts() {
        occCounts();
        vector<array<int, 3>> res;
        for (int v = 2; v < (int)palLen.size(); v++) {
            int e = firstEnd[v];
            res.push_back({e - palLen[v] + 1, e, occCache[v]});
        }
        return res;
    }

    // O(n log n) -- fewest palindromes a can be cut into
    int minPalPartition() {
        const int INF = INT_MAX / 2;
        vector<int> dp(n + 1, INF), seriesAns(palLen.size(), 0);
        dp[0] = 0;
        for (int i = 1; i <= n; i++) {
            for (int v = sufNode[i - 1]; palLen[v] > 0; v = sLink[v]) {
                seriesAns[v] = dp[i - palLen[sLink[v]] - diff[v]];
                if (diff[v] == diff[palLink[v]])
                    seriesAns[v] = min(seriesAns[v], seriesAns[palLink[v]]);
                dp[i] = min(dp[i], seriesAns[v] + 1);
            }
        }
        return dp[n];
    }

    vector<int> &occCounts() {
        if (occCache.empty()) {
            occCache.assign(palLen.size(), 0);
            for (int v : sufNode) occCache[v]++;
            for (int v = (int)palLen.size() - 1; v >= 2; v--)
                occCache[palLink[v]] += occCache[v];
            occCache[0] = occCache[1] = 0;
        }
        return occCache;
    }
};